crates/widgets/src/tab_widget.rs

```diff
@@ -0,0 +1,382 @@
+    {
+        //Create the new tab
+        let header = create_tab_header(ctx,header_text,body);
+
+        //Set tab body hidden
+        ctx.get_widget(body).set("visibility",Visibility::Hidden);
+
+        //Push button to the header container
+        ctx.append_child_entity_to(header,self.header_container);
+
+        //Push the body to the body container
+        ctx.append_child_entity_to(body,self.body_container);
+
+        //Push the new tab to the list
+        self.tabs.push((header,body));
+
+        //If the added tab is the first
+        if self.tabs.len() == 1
+        {
+            //Select the tab just inserted
+            self.selected = 0;
+            self.refresh_selected_tab(ctx);
+        }
+    }
+
+    fn remove_tab_internal(&mut self,ctx: &mut Context,body: Entity)
+    {
+        match self.get_index(body)
+        {
+            Some(index)=>
+            {
+                let (header,body) = self.tabs.remove(index);
+
+                //Push button to the header container
+                ctx.remove_child_from(header,self.header_container);
+
+                //Push the body to the body container
+                ctx.remove_child_from(body,self.body_container);
+
+
+                //If there is at least one tab
+                if self.tabs.len() != 0
+                {
+                    //If selected is greater than tab count, select the last one
+                    if self.selected >= self.tabs.len(){self.selected = self.tabs.len()-1;}
+
+                    //Only update current selection if the removed tab is lesser than the selected.
+                    //If it is greater, there is no need to update, but simply remove the target tab
+                    if index <= self.selected {self.refresh_selected_tab(ctx);}
+                }
+
+            }
+            None=>()
+        }
+    }
+
+}
+
+impl State for TabWidgetState {
+    fn init(&mut self, _: &mut Registry, ctx: &mut Context)
+    {
+        self.header_container = ctx.child(HEADER_CONTAINER).entity();
+        self.body_container = ctx.child(BODY_CONTAINER).entity();
+    }
+    fn update(&mut self, _: &mut Registry, ctx: &mut Context)
+    {
+        let actions: Vec<Action> = self.actions.drain(..).collect();
+        for action in actions
+        {
+            match action
+            {
+                Action::SelectByIndex(index)=>{self.select_by_index_internal(ctx,index);}
+                Action::SelectByBody(body)=>
+                {
+                    if let Some(index) = self.get_index(body){self.select_by_index_internal(ctx,index)}
+                }
+                Action::Add(header_text,body)=>{self.add_tab_internal(ctx,header_text,body);}
+                Action::Remove(body)=>{self.remove_tab_internal(ctx,body);}
+            }
+        }
+    }
+}
+
+widget!(
+    /// The `TabWidget` widget can store and control multiple tabs with arbitrary content. Only the selected tab will show it's content.
```

Comment:
Could you add a documentation example please?

```rust
/// The `TabWidget` widget can store and control multiple tabs with arbitrary content. Only the selected tab will show it's content.
///
/// This example creates a TabWidget:
/// ```rust
///   TabWidget::new()
///       .tab("Tab header 1",TextBlock::new().text("Tab content 1").build(ctx))
///       .tab("Tab header 2",TextBlock::new().text("Tab content 2").build(ctx))
///       .tab("Tab header 3",TextBlock::new().text("Tab content 3").build(ctx))
///       .build(ctx)
/// ```
```